Update the Fixer.Io api endpoint and method of submitting the api key.

richard-parnaby-king opened the pull request:
### Description (*)
Fixer.io has changed the api endpoint and method of submitting the api key. Utilising Fixer.io to update currency rates results in an invalid api key error message. This PR updates the api endpoint and method of submitting the api key.

### Manual testing scenarios (*)
Go to Stores > Currency Rates , select Fixer.io as service and click on "Import". An error will appear.

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#36010


### Contribution checklist (*)
- [x]  Pull request has a meaningful description of its purpose
- [x]  All commits are accompanied by meaningful commit messages
- [x]  All new or changed code is covered with unit/integration tests (if applicable)
- [ ]  README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
- [ ]  All automated tests passed successfully (all builds are green)

richard-parnaby-king committed:
Fixer.Io api change fix

The url to connect to has changed, and the api key is now sent in the headers as per the documentation:
https://apilayer.com/marketplace/fixer-api?live_demo=show#authentication

richard-parnaby-king commented:
@magento run all tests

richard-parnaby-king committed:
Reduce line character length

richard-parnaby-king commented:
@magento run Static Tests

richard-parnaby-king committed:
Merge branch '2.4-develop' into patch-1

ihor-sviziev commented:
Hi @richard-parnaby-king,
Thank you for your contribution!
I'm closing this PR as a duplicate of https://github.com/magento/magento2/pull/35984 that was created earlier

richard-parnaby-king commented:
@ihor-sviziev That pull request does not solve the issue. The fixer api requires the access key to be submitted as part of a custom header, not as part of the url query string, which is what this pull request resolves.

ihor-sviziev commented on app/code/Magento/Directory/Model/Currency/Import/FixerIo.php:
```diff
@@ -23,8 +23,13 @@ class FixerIo extends AbstractImport
+    /**
+     * @var string
+     */
+    public const API_KEY_CONFIG_PATH = 'currency/fixerio/api_key';
```
Let's make it private due to backward compatibility reasons
```suggestion
    private const API_KEY_CONFIG_PATH = 'currency/fixerio/api_key';
```

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (changes requested):
@richard-parnaby-king please review and fix failing tests

Den4ik reviewed (changes requested):
I believe that use of access key should be also removed at lines 114-115

```
url = str_replace(
            ['{{ACCESS_KEY}}', '{{CURRENCY_FROM}}', '{{CURRENCY_TO}}'],
            [$accessKey, $currencyFrom, $currenciesStr],
            self::CURRENCY_CONVERTER_URL
);
```

richard-parnaby-king commented:
@ihor-sviziev

Some of these tests do not revolve around the functionality of this pull request, for example there is a failing test for the Aws S3 remote filestorage component. 

Looking at the failing tests the only one concerning the FixerIo module is failing this test: testFetchRates (https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Directory/Test/Unit/Model/Currency/Import/FixerIoTest.php)

The error message is:
```
Expectation failed for method name is "getValue" when invoked zero or more times
Parameter 0 for invocation #1 Magento\Framework\App\Config\ScopeConfigInterface::getValue('currency/fixerio/api_key', 'store', null) does not match expected value.
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'currency/fixerio/timeout'
+'currency/fixerio/api_key'
```

Which is this bit of code:
```
        $this->scopeConfig->method('getValue')
            ->withConsecutive(
                ['currency/fixerio/api_key', 'store'],
                ['currency/fixerio/timeout', 'store']
            )
            ->willReturnOnConsecutiveCalls('api_key', 100);
```

Do I need to fix this or does it sit with the core team? If I need to fix it, what is the expected functionality?

richard-parnaby-king committed:
Merge branch '2.4-develop' into patch-1

richard-parnaby-king committed:
Update app/code/Magento/Directory/Model/Currency/Import/FixerIo.php

Co-authored-by: Ihor Sviziev <ihor-sviziev@users.noreply.github.com>

ihor-sviziev committed:
Merge branch '2.4-develop' into patch-1

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@richard-parnaby-king, please fix the related tests only. For instance, the one you said fails because of changes.

richard-parnaby-king committed:
Removed ACCESS_KEY param from url str_replace.

Removed per recommendation from Den4ik (https://github.com/magento/magento2/pull/36077#pullrequestreview-1096486845)

richard-parnaby-king committed:
Unit Test updated to reflect changes to code

`$message` updated with new api endpoint.
scopeConfig->getValue called multiple times. Updated `withConsecutive` calls.
Added `$httpClient->method('getHeaders')` return type.  Should be `Laminas\Http\Headers` but set as `DataObject` as Headers is not included in Magento.

richard-parnaby-king committed:
Merge branch '2.4-develop' into patch-1

richard-parnaby-king commented:
@magento run Unit Tests

richard-parnaby-king committed:
Removed getHeaders mock call.

`Trying to configure method "getHeaders" which cannot be configured because it does not exist, has not been specified, is final, or is static` so removed function call.

richard-parnaby-king commented:
https://github.com/magento run Unit Tests

richard-parnaby-king committed:
Readded getHeaders mock.

Misunderstood error message. Re-added getHeaders mock call. Added getHeaders call to mock of LaminasClient (`$httpClient`) so PHPUnit knows it's a callable function.

richard-parnaby-king commented:
https://github.com/magento run Unit Tests

richard-parnaby-king commented:
@magento run Unit Tests

richard-parnaby-king committed:
Merge branch '2.4-develop' into patch-1

richard-parnaby-king commented:
@magento run all tests

richard-parnaby-king committed:
Merge branch '2.4-develop' into patch-1

richard-parnaby-king commented:
@magento run All Tests

richard-parnaby-king commented:
@ihor-sviziev Failing tests are unrelated to this PR.

ihor-sviziev commented:
@magento run Functional Tests EE, WebAPI Tests

engcom-Lima commented:
@magento run Functional Tests EE

engcom-Lima commented:
@magento give me 2.4-develop instance

engcom-Lima commented:
Hi @richard-parnaby-king 

Thanks for your contribution and collaboration.

I have tried to reproduce the issue ,it is reproducible to me below is the screenshot for your reference:
<img width="1512" alt="Screenshot 2022-09-15 at 5 39 06 PM" src="https://user-images.githubusercontent.com/60198022/190413960-400fbb57-9282-4073-a91a-5b070589f993.png">

But after applying PR changes I am getting error in UI as well as in logs as below:
**IN LOGS:**
```
[2022-09-15T11:51:32.612615+00:00] main.CRITICAL: Error: Call to undefined method Magento\Framework\HTTP\LaminasClient::getHeaders() in /opt/homebrew/var/www/magento2/app/code/Magento/Directory/Model/Currency/Import/FixerIo.php:176
Stack trace:
#0 /opt/homebrew/var/www/magento2/app/code/Magento/Directory/Model/Currency/Import/FixerIo.php(121): Magento\Directory\Model\Currency\Import\FixerIo->getServiceResponse('https://api.api...')
#1 /opt/homebrew/var/www/magento2/app/code/Magento/Directory/Model/Currency/Import/FixerIo.php(81): Magento\Directory\Model\Currency\Import\FixerIo->convertBatch(Array, 'INR', Array)
#2 /opt/homebrew/var/www/magento2/app/code/Magento/CurrencySymbol/Controller/Adminhtml/System/Currency/FetchRates.php(87): Magento\Directory\Model\Currency\Import\FixerIo->fetchRates()
#3 /opt/homebrew/var/www/magento2/lib/internal/Magento/Framework/Interception/Interceptor.php(58): Magento\CurrencySymbol\Controller\Adminhtml\System\Currency\FetchRates->execute()
```
**IN UI:**
<img width="1512" alt="Screenshot 2022-09-15 at 6 23 26 PM" src="https://user-images.githubusercontent.com/60198022/190414921-2bb9c7f6-fec3-477b-95bd-1ce6e319b0be.png">

I have tested it in local as well in git cloud instance as well but issue is same.

Since PR is not working fine. I am moving this PR to **Changes requested**.

Thanks

engcom-Lima reviewed (changes requested):
Hi @richard-parnaby-king 

Thanks for your contribution and collaboration.

I have tried to reproduce the issue ,it is reproducible to me below is the screenshot for your reference:
<img width="1512" alt="Screenshot 2022-09-15 at 5 39 06 PM" src="https://user-images.githubusercontent.com/60198022/190413960-400fbb57-9282-4073-a91a-5b070589f993.png">

But after applying PR changes I am getting error in UI as well as in logs as below:
**IN LOGS:**
```
[2022-09-15T11:51:32.612615+00:00] main.CRITICAL: Error: Call to undefined method Magento\Framework\HTTP\LaminasClient::getHeaders() in /opt/homebrew/var/www/magento2/app/code/Magento/Directory/Model/Currency/Import/FixerIo.php:176
Stack trace:
#0 /opt/homebrew/var/www/magento2/app/code/Magento/Directory/Model/Currency/Import/FixerIo.php(121): Magento\Directory\Model\Currency\Import\FixerIo->getServiceResponse('https://api.api...')
#1 /opt/homebrew/var/www/magento2/app/code/Magento/Directory/Model/Currency/Import/FixerIo.php(81): Magento\Directory\Model\Currency\Import\FixerIo->convertBatch(Array, 'INR', Array)
#2 /opt/homebrew/var/www/magento2/app/code/Magento/CurrencySymbol/Controller/Adminhtml/System/Currency/FetchRates.php(87): Magento\Directory\Model\Currency\Import\FixerIo->fetchRates()
#3 /opt/homebrew/var/www/magento2/lib/internal/Magento/Framework/Interception/Interceptor.php(58): Magento\CurrencySymbol\Controller\Adminhtml\System\Currency\FetchRates->execute()
```
**IN UI:**
<img width="1512" alt="Screenshot 2022-09-15 at 6 23 26 PM" src="https://user-images.githubusercontent.com/60198022/190414921-2bb9c7f6-fec3-477b-95bd-1ce6e319b0be.png">

I have tested it in local as well in git cloud instance as well but issue is same.

Since PR is not working fine. I am moving this PR to **Changes requested**.

Thanks

richard-parnaby-king committed:
LaminasClient::getHeaders does not exist. Refactored.

richard-parnaby-king committed:
Update test.

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented on app/code/Magento/Directory/Test/Unit/Model/Currency/Import/FixerIoTest.php:
```diff
@@ -110,6 +111,8 @@ public function testFetchRates(): void
             ->willReturnSelf();
         $httpClient->method('setMethod')
             ->willReturnSelf();
+        $httpClient->method('getHeaders')
+            ->willReturnSelf();
```
Maybe we should check if method setHeaders was executed with correct params?

richard-parnaby-king commented:
@magento give me test instance

richard-parnaby-king committed:
JSON Decode response from remote api.

richard-parnaby-king committed:
Made validateResponse if statement more robust.

richard-parnaby-king commented:
@magento give me test instance

richard-parnaby-king commented:
@magento give me test instance

richard-parnaby-king committed:
Var export $response for debugging,

richard-parnaby-king commented:
@magento give me test instance

richard-parnaby-king commented:
@magento give me test instance

ishakhsuvarov commented:
Looks like requested changes were not implemented or discussed in a long time, hence I am closing this PR for now.

Please feel free to reopen if there is any additional work done on this.
